src/hardware/bk-1856d/protocol.c

```diff
@@ -0,0 +1,375 @@
+			continue;
+		else if (devc->buffer[idx] == '0' && zero_found == 0)
+		{
+			zero_found = 1;
+			continue;
+		}
+		else
+		{
+			has_data = 1;
+			break;
+		}
+	}
+	return !has_data;
+}
+
+static void bk_1856d_send_packet(
+	const struct sr_dev_inst *sdi, double freq_value, int digits)
+{
+	struct sr_datafeed_packet packet;
+	struct sr_datafeed_analog analog;
+	struct sr_analog_encoding encoding;
+	struct sr_analog_meaning meaning;
+	struct sr_analog_spec spec;
+
+	sr_analog_init(&analog, &encoding, &meaning, &spec, digits);
+	analog.meaning->mq = SR_MQ_FREQUENCY;
+	analog.meaning->unit = SR_UNIT_HERTZ;
+	analog.meaning->channels = sdi->channels;
+	analog.num_samples = 1;
+	analog.data = &freq_value;
+	analog.encoding->unitsize = sizeof(freq_value);
+	analog.encoding->is_float = TRUE;
+	analog.encoding->digits = digits;
+
+	packet.type = SR_DF_ANALOG;
+	packet.payload = &analog;
+	sr_session_send(sdi, &packet);
+}
+
+static void bk_1856d_parse_message(const struct sr_dev_inst *sdi)
+{
+	struct dev_context *devc;
+	struct sr_serial_dev_inst *serial;
+
+	double freq_value;
+	int digits;
+	char *endPtr, *dotPtr;
+
+	if (!(devc = sdi->priv) || !(serial = sdi->conn))
+		return;
+
+	/* check for cr at end of message */
+	if (devc->buffer[BK1856D_MSG_SIZE-1] != '\xD')
+	{
+		sr_err("expected cr at end of message.");
+		devc->buffer_level = 0;
+		serial_flush(serial);
+		g_mutex_lock(&devc->rw_mutex);
+		bk_1856d_send_input_sel(sdi);
+		bk_1856d_send_gate_time(sdi);
+		bk_1856d_request_data(sdi);
+		g_mutex_unlock(&devc->rw_mutex);
+		return;
+	}
+
+	devc->buffer[BK1856D_MSG_SIZE-1] = 0; /* set trailing zero */
+
+	if (bk_1856d_check_for_zero_message(devc))
+	{
+		sr_spew("received 0 package");
+		devc->buffer_level = 0;
+		bk_1856d_request_data(sdi);
+		return;
+	}
+
+	sr_dbg("received msg: '%s'", devc->buffer);
+	freq_value = strtod(devc->buffer, &endPtr);
+	sr_dbg("parsed value: %f", freq_value);
+
+	if (strcmp(devc->buffer + BK1856D_MSG_NUMBER_SIZE+1, "Hz ") != 0)
+	{
+		sr_err("not a frequency returned");
+		devc->buffer_level = 0;
+		g_mutex_lock(&devc->rw_mutex);
+		bk_1856d_send_input_sel(sdi);
+		bk_1856d_send_gate_time(sdi);
+		bk_1856d_request_data(sdi);
+		g_mutex_unlock(&devc->rw_mutex);
+		return;
+	}
+
+	dotPtr = strchr(devc->buffer, '.');
+	if (dotPtr)
+		digits = endPtr - (dotPtr+1);
+	else
+		digits = endPtr - devc->buffer;
+
+	if (devc->buffer[BK1856D_MSG_NUMBER_SIZE] == 'M') {
+		freq_value *= 1e6;
+		digits -= 6;
+	}
+	if (devc->buffer[BK1856D_MSG_NUMBER_SIZE] == 'k') {
+		freq_value *= 1e3;
+		digits -= 3;
+	}
+
+	bk_1856d_send_packet(sdi, freq_value, digits);
+
+	sr_sw_limits_update_samples_read(&(devc->sw_limits), 1);
+
+	if (!sr_sw_limits_check(&(devc->sw_limits)))
+	{
+		devc->buffer_level = 0;
+		g_mutex_lock(&devc->rw_mutex);
+		bk_1856d_chk_select_input(sdi);
+		bk_1856d_send_gate_time(sdi);
+		bk_1856d_request_data(sdi);
+		g_mutex_unlock(&devc->rw_mutex);
+	}
+	else
+		sr_dev_acquisition_stop((struct sr_dev_inst *)sdi);
```

Comment:
you are explicitly casting to bypass the 'const' attribute, this never feels right. I see other drivers also do this, would like to hear input from others about this.

Reply:
This shouldn't be cast because it shouldn't be `const` in the first place. The mistake is in the next function.